### ISSUE TITLE
Joplin 3 - Different required fields for published/draft

### ISSUE THREAD

**Issue by briaguya** [cityofaustin/techstack#2767]
I think this is something we could accomplish by using a custom publish view/template.

This has a lot of potential, and I have a 
## couple of ideas:
Service Page:
- i1 - Topics, we could have this be a text field for drafts [t1]
  -  i1a - when we get to the publish screen, we could have a user flow for "you have something in draft status that needs to be updated before you can publish,"
    - i1ai - that could even include a "create new topic" flow if the topic is just text at that point.

## General tangential braindump:
- t1: or maybe an autocomplete, I'd like to investigate wagtail autocomplete a bit
- t2: all one page type, just modify react-streamfield to show different options at different points, it might flow kinda like a wizard at first but with some modification there's a lot of potential here.
- t3: need to investigate how to do required fields within a streamfield, can you make a streamfield where you need to add a certain type of block? this could make it so we can't save a draft of a "page" until we have a name and page type
- t4: we can make a lot of different custom blocks, step blocks in "steps" blocks, figure out custom logic to only allow certain ones to show up, there's a lot of potential here

## Somehow pulling it together into a proposed opportunity to go exploring:
I'd like to investigate the possibility of using a single page type with react-streamfield, and implementing custom logic based on other fields in the stream.
This could look like:
* Everything is a snippet, more specifically, one kind of snippet
   * There is one field on this snippet
      * It is a StreamField [t3]
* DRAFTS
  * For any relations (related topics), we allow either text or the relation
    * should look into wagtail-autocomplete for this
  * Custom logic in a modified react-streamfield could be used to determine:
    * If we've entered enough data to save a draft
    * What blocks [t4] to make available based on page type
* CUSTOM PUBLISH VIEW
  * basically we just render the edit view again, but in "publish mode"
    * we pass a different set of requirements into react-streamfield, 
      * for example, make sure our autocomplete boxes have things selected, if they don't [i1ai]
  * See if we can send a publishable flag in there
* Maybe not even have it as a custom view, but instead override the hooks for the publish button itself

**Comment by easherma** [comment]
Cool ideas! I think we can probably try something simpler by extending the modelform to do some custom validation: http://docs.wagtail.io/en/v2.6.1/advanced_topics/customisation/page_editing_interface.html#customising-generated-forms

So we could probably do some validation checks that vary between drafts vs publishing

**Comment by briaguya** [comment]
@easherma I think modifying our current extended modelforms to handle that could be great! It's probably worth making an issue for it and seeing if we want to get that implemented in current Joplin. By doing something as out there as extending react-streamfield it opens up opportunities to handle more complex UX changes on a field by field basis, opening doors to handling the rest of the iceberg that we're just hitting the tip of so to speak.

**Comment by desigonz** [comment]
related to #2180

**Comment by easherma** [comment]
Also related to https://github.com/cityofaustin/techstack/issues/2732

I'll comment here too:
Idea 1:
Make every field blank=True so the draft will save. But then on publish (maybe in the save() method) do long-hand form validation, and raise exception and add errors to missing fields?

Idea 2:
There’s probably a better way using base_form_class http://docs.wagtail.io/en/v2.6.1/advanced_topics/customisation/page_editing_interface.html#customising-generated-forms

Maybe make an intermediary class that stores draft data, and on-page-publish it actually saves the page?
